connmgr/connmanager.go

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) 2016 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package connmgr
+
+import (
+	"errors"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// maxFailedAttempts is the maximum number of successive failed connection
+// attempts after which network failure is assumed and new connections will
+// be delayed by the configured retry duration.
+const maxFailedAttempts = 10
+
+var (
+	//ErrDialNil is used to indicate that Dial cannot be nil in the configuration.
+	ErrDialNil = errors.New("Config: Dial cannot be nil")
+
+	// maxRetryDuration is the max duration of time retrying of a persistent
+	// connection is allowed to grow to.  This is necessary since the retry
+	// logic uses a backoff mechanism which increases the interval base times
+	// the number of retries that have been done.
+	maxRetryDuration = time.Minute * 5
+
+	// defaultRetryDuration is the default duration of time for retrying
+	// persistent connections.
+	defaultRetryDuration = time.Second * 5
+
+	// defaultMaxOutbound is the default number of maximum outbound connections
+	// to maintain.
+	defaultMaxOutbound = uint32(8)
+)
+
+// DialFunc defines a function that dials a connection.
+type DialFunc func(string, string) (net.Conn, error)
+
+// AddressFunc defines a function that returns a network address to connect to.
+type AddressFunc func() string
+
+// ConnState represents the state of the requested connection.
+type ConnState uint8
+
+// OnConnectionFunc is the signature of the callback function which is used to
+// subscribe to new connections.
+type OnConnectionFunc func(*ConnReq, net.Conn)
+
+// OnDisconnectionFunc is the signature of the callback function which is used to
+// notify disconnections.
+type OnDisconnectionFunc func(*ConnReq)
+
+// ConnState can be either pending, established, disconnected or failed.  When
+// a new connection is requested, it is attempted and categorized as
+// established or failed depending on the connection result.  An established
+// connection which was disconnected is categorized as disconnected.
+const (
```

Comment:
Please move this up to just after the type definition for `ConnState` so they're grouped.